src/main/java/de/tum/in/www1/artemis/config/migration/MigrationService.java

```diff
@@ -0,0 +1,158 @@
+package de.tum.in.www1.artemis.config.migration;
+
+import static tech.jhipster.config.JHipsterConstants.SPRING_PROFILE_TEST;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+import org.apache.commons.codec.binary.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.core.env.Profiles;
+import org.springframework.stereotype.Service;
+import org.thymeleaf.util.StringUtils;
+
+import de.tum.in.www1.artemis.domain.MigrationChangelog;
+import de.tum.in.www1.artemis.repository.MigrationChangeRepository;
+
+/**
+ * This service contains utility functionality that verifies a changelog to prevent corruption and executes a given changelog.
+ */
+@Service
+public class MigrationService {
+
+    private final Logger log = LoggerFactory.getLogger(MigrationRegistry.class);
+
+    @Value("${artemis.version}")
+    private String artemisVersion;
+
+    private final MessageDigest messageDigest = MessageDigest.getInstance("MD5");
+
+    private final MigrationChangeRepository migrationChangeRepository;
+
+    private final AutowireCapableBeanFactory beanFactory;
+
+    public MigrationService(MigrationChangeRepository migrationChangeRepository, AutowireCapableBeanFactory beanFactory) throws NoSuchAlgorithmException {
+        this.migrationChangeRepository = migrationChangeRepository;
+        this.beanFactory = beanFactory;
+    }
+
+    /**
+     * First checks the integrity of the passed changelog, then executes not yet executed entries.
+     * After each execution it marks the entry as executed. All entries of one startup run get the same hash assigned.
+     *
+     * @param event         Specifies when this method gets called and provides the event with all application data
+     * @param entryClassMap The changelog to be executed
+     */
+    public void execute(ApplicationReadyEvent event, SortedMap<Integer, Class<? extends MigrationEntry>> entryClassMap) {
+        if (event.getApplicationContext().getEnvironment().acceptsProfiles(Profiles.of(SPRING_PROFILE_TEST))) {
+            return;
+        }
+
+        log.info("Starting Artemis migration");
+
+        SortedMap<Integer, MigrationEntry> entryMap = instantiateEntryMap(entryClassMap);
+
+        if (!checkIntegrity(entryMap)) {
+            log.error("{} corrupted. Aborting startup.", getClass().getSimpleName());
+            event.getApplicationContext().close();
+            System.exit(1);
```

Comment:
Could this lead to problems when deploying the Artemis WAR on e.g. Tomcat? This would terminate the whole JVM instead of just Artemis.
Closing the application context is probably enough to shutdown Artemis itself.

Reply:
I used the exit because I got exceptions that I couldn't get rid of. I wasn't aware that it can cause such issues. I decided to go with my own exception to prevent unrelated exceptions.

Reply:
Thanks, that looks good. I added a separate suggestion about the exception in another review comment.